Adopt dynamicDowncast<>() more in html/

cdumez opened the pull request:
#### 7ccc0ce3e7a39a98dee3312d14472abdf92d2fcc
<pre>
Adopt dynamicDowncast&lt;&gt;() more in html/
<a href="https://bugs.webkit.org/show_bug.cgi?id=265174">https://bugs.webkit.org/show_bug.cgi?id=265174</a>

Reviewed by Jean-Yves Avenard and Tim Nguyen.

* Source/WebCore/html/HTMLElement.h:
(WebCore::Node::hasTagName const):
(isType):
* Source/WebCore/html/HTMLEmbedElement.cpp:
(WebCore::findWidgetRenderer):
(WebCore::HTMLEmbedElement::rendererIsNeeded):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::autofillMantle const):
* Source/WebCore/html/HTMLFormControlElement.h:
(isType):
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::submitImplicitly):
(WebCore::HTMLFormElement::textFieldValues const):
(WebCore::HTMLFormElement::findSubmitButton):
(WebCore::HTMLFormElement::registerFormListedElement):
(WebCore::HTMLFormElement::findSubmitter const):
(WebCore::HTMLFormElement::defaultButton const):
(WebCore::HTMLFormElement::constructEntryList):
* Source/WebCore/html/HTMLFrameElementBase.h:
(isType):
* Source/WebCore/html/HTMLFrameOwnerElement.cpp:
(WebCore::HTMLFrameOwnerElement::renderWidget const):
* Source/WebCore/html/HTMLFrameSetElement.cpp:
(WebCore::HTMLFrameSetElement::namedItem):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::setBestFitURLAndDPRFromImageCandidate):
(WebCore::HTMLImageElement::bestFitSourceFromPictureElement):
(WebCore::HTMLImageElement::attributeChanged):
(WebCore::HTMLImageElement::didAttachRenderers):
(WebCore::HTMLImageElement::insertedIntoAncestor):
(WebCore::HTMLImageElement::isSystemPreviewImage const):
(WebCore::HTMLImageElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLImageLoader.cpp:
(WebCore::HTMLImageLoader::notifyFinished):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::willDispatchEvent):
* Source/WebCore/html/HTMLLIElement.cpp:
(WebCore::HTMLLIElement::didAttachRenderers):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::isEventTargetedAtInteractiveDescendants const):
(WebCore::HTMLLabelElement::defaultEventHandler):
* Source/WebCore/html/HTMLLegendElement.cpp:
(WebCore::HTMLLegendElement::form const):
* Source/WebCore/html/HTMLMaybeFormAssociatedCustomElement.h:
(isType):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::bestMediaElementForRemoteControls):
(WebCore::HTMLMediaElement::updateActiveTextTrackCues):
(WebCore::HTMLMediaElement::textTrackRemoveCue):
(WebCore::HTMLMediaElement::setReadyState):
(WebCore::HTMLMediaElement::mediaPlayerSizeChanged):
(WebCore::HTMLMediaElement::mediaPlayerRenderingCanBeAccelerated):
(WebCore::HTMLMediaElement::enterFullscreen):
(WebCore::HTMLMediaElement::exitFullscreen):
(WebCore::HTMLMediaElement::willBecomeFullscreenElement):
(WebCore::HTMLMediaElement::mediaState const):
(WebCore::HTMLMediaElement::isVideoTooSmallForInlinePlayback):
* Source/WebCore/html/HTMLMediaElement.h:
(isType):
* Source/WebCore/html/HTMLMeterElement.cpp:
(WebCore::HTMLMeterElement::renderMeter const):
* Source/WebCore/html/HTMLNameCollection.cpp:
(WebCore::isObjectElementForDocumentNameCollection):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::preventsParentObjectFromExposure):
(WebCore::HTMLObjectElement::updateExposedState):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::ownerSelectElement const):
(WebCore::HTMLOptionElement::isDisabledFormControl const):
(WebCore::HTMLOptionElement::collectOptionInnerText const):
* Source/WebCore/html/HTMLPictureElement.cpp:
(WebCore::HTMLPictureElement::isSystemPreviewImage):
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::defaultEventHandler):
(WebCore::HTMLPlugInElement::supportsFocus const):
(WebCore::HTMLPlugInElement::setReplacement):
(WebCore::HTMLPlugInElement::isReplacementObscured):
* Source/WebCore/html/HTMLPlugInImageElement.cpp:
(WebCore::HTMLPlugInImageElement::didAttachRenderers):
* Source/WebCore/html/HTMLPlugInImageElement.h:
(isType):
* Source/WebCore/html/HTMLProgressElement.cpp:
(WebCore::HTMLProgressElement::renderProgress const):
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::optionToSelectFromChildChangeScope):
(WebCore::HTMLSelectElement::setLength):
(WebCore::HTMLSelectElement::nextSelectableListIndexPageAway const):
(WebCore::HTMLSelectElement::saveLastSelection):
(WebCore::HTMLSelectElement::setActiveSelectionAnchorIndex):
(WebCore::HTMLSelectElement::updateListBoxSelection):
(WebCore::HTMLSelectElement::listBoxOnChange):
(WebCore::HTMLSelectElement::scrollToSelection):
(WebCore::HTMLSelectElement::setOptionsChangedOnRenderer):
(WebCore::HTMLSelectElement::recalcListItems const):
(WebCore::HTMLSelectElement::selectedIndex const):
(WebCore::HTMLSelectElement::selectOption):
(WebCore::HTMLSelectElement::deselectItemsWithoutValidation):
(WebCore::HTMLSelectElement::saveFormControlState const):
(WebCore::HTMLSelectElement::searchOptionsForValue const):
(WebCore::HTMLSelectElement::restoreFormControlState):
(WebCore::HTMLSelectElement::appendFormData):
(WebCore::HTMLSelectElement::reset):
(WebCore::HTMLSelectElement::platformHandleKeydownEvent):
(WebCore::HTMLSelectElement::menuListDefaultEventHandler):
(WebCore::HTMLSelectElement::updateSelectedState):
(WebCore::HTMLSelectElement::listBoxDefaultEventHandler):
(WebCore::HTMLSelectElement::defaultEventHandler):
(WebCore::HTMLSelectElement::lastSelectedListIndex const):
(WebCore::HTMLSelectElement::optionAtIndex const):
(WebCore::HTMLSelectElement::accessKeySetSelectedIndex):

Canonical link: <a href="https://commits.webkit.org/271033@main">https://commits.webkit.org/271033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3832a69e2f5e0f51743edfd4127a2fca3e8084f5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27136 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5752 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24815 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27594 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24655 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27398 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4577 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23292 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3990 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4096 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24289 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29979 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24779 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24703 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30278 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26316 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4104 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2290 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28199 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5565 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33772 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4564 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7318 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3509 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4480 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -195,13 +195,14 @@ void HTMLFormElement::submitImplicitly(Event& event, bool fromImplicitSubmission
     unsigned submissionTriggerCount = 0;
     for (auto& listedElement : m_listedElements) {
-        if (!is<HTMLFormControlElement>(*listedElement))
+        auto* formElement = dynamicDowncast<HTMLFormControlElement>(*listedElement);
+        if (!formElement)
             continue;
```
I have a slight preference for wrapping the whole block with
`if (auto* formElement = dynamicDowncast<HTMLFormControlElement>(*listedElement))` rather than early `continue;` here.

cdumez replied:
In this instance, I prefer to keep it this way as it matches more the WebKit coding style where we try to early return to minimize nesting.

I often make an exception for one-liners but here the body is larger than that.

nt1m commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -338,13 +339,12 @@ ExceptionOr<void> HTMLFormElement::requestSubmit(HTMLElement* submitter)
-        auto& input = downcast<HTMLInputElement>(*element);
-        if (!input.isTextField())
+        if (!input->isTextField())
             return std::nullopt;
```
Combine these 2 if statements ?

`if (!input || !input->isTextField())`

cdumez replied:
Ok

nt1m commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -356,13 +356,13 @@ RefPtr<HTMLFormControlElement> HTMLFormElement::findSubmitButton(HTMLFormControl
     RefPtr<HTMLFormControlElement> firstSuccessfulSubmitButton;
     for (auto& listedElement : m_listedElements) {
-        if (!is<HTMLFormControlElement>(*listedElement))
+        auto* control = dynamicDowncast<HTMLFormControlElement>(*listedElement);
+        if (!control)
             continue;
```
Ditto, I have a slight preference with wrapping the whole block with `if (auto* control = dynamicDowncast<HTMLFormControlElement>(*listedElement))` instead of early `continue`.

cdumez replied:
Same here, the body is too large, I'd rather reduce nesting.

nt1m commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -748,10 +745,14 @@ bool HTMLFormElement::wasUserSubmitted() const
+        return nullptr;
+    auto* node = dynamicDowncast<Node>(event->target());
+    if (!node)
         return nullptr;
-    auto& node = downcast<Node>(*event->target());
-    auto* element = is<Element>(node) ? &downcast<Element>(node) : node.parentElement();
+    auto* element = dynamicDowncast<Element>(*node);
```
Can we dynamicDowncast directly to Element?

cdumez replied:
No, it would break the code. This code needs to be able to distinguish Node vs Element:
1. If not a Node, early return
2. If an Element, use that element
3. If an Node but not an element, use that node's parentElement.

nt1m replied:
Ah right, I missed the following lines

nt1m commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -760,12 +761,12 @@ HTMLFormControlElement* HTMLFormElement::defaultButton() const
-        if (control.isSuccessfulSubmitButton()) {
-            m_defaultButton = control;
-            return &control;
+        if (control->isSuccessfulSubmitButton()) {
```
`if (auto* control = dynamicDowncast<HTMLFormControlElement>(*listedElement); control && control->isSuccessfulSubmitButton())`

cdumez replied:
Ok

jyavenard commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -8664,7 +8668,7 @@ MediaProducerMediaStateFlags HTMLMediaElement::mediaState() const
     bool streaming = false;
 #if ENABLE(MANAGED_MEDIA_SOURCE)
-    RefPtr managedMediasource = is<ManagedMediaSource>(m_mediaSource) ? downcast<ManagedMediaSource>(m_mediaSource.get()) : nullptr;
+    RefPtr managedMediasource = dynamicDowncast<ManagedMediaSource>(m_mediaSource);
```
nice

cdumez committed:
Adopt dynamicDowncast<>() more in html/
https://bugs.webkit.org/show_bug.cgi?id=265174

Reviewed by Jean-Yves Avenard and Tim Nguyen.

* Source/WebCore/html/HTMLElement.h:
(WebCore::Node::hasTagName const):
(isType):
* Source/WebCore/html/HTMLEmbedElement.cpp:
(WebCore::findWidgetRenderer):
(WebCore::HTMLEmbedElement::rendererIsNeeded):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::autofillMantle const):
* Source/WebCore/html/HTMLFormControlElement.h:
(isType):
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::submitImplicitly):
(WebCore::HTMLFormElement::textFieldValues const):
(WebCore::HTMLFormElement::findSubmitButton):
(WebCore::HTMLFormElement::registerFormListedElement):
(WebCore::HTMLFormElement::findSubmitter const):
(WebCore::HTMLFormElement::defaultButton const):
(WebCore::HTMLFormElement::constructEntryList):
* Source/WebCore/html/HTMLFrameElementBase.h:
(isType):
* Source/WebCore/html/HTMLFrameOwnerElement.cpp:
(WebCore::HTMLFrameOwnerElement::renderWidget const):
* Source/WebCore/html/HTMLFrameSetElement.cpp:
(WebCore::HTMLFrameSetElement::namedItem):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::setBestFitURLAndDPRFromImageCandidate):
(WebCore::HTMLImageElement::bestFitSourceFromPictureElement):
(WebCore::HTMLImageElement::attributeChanged):
(WebCore::HTMLImageElement::didAttachRenderers):
(WebCore::HTMLImageElement::insertedIntoAncestor):
(WebCore::HTMLImageElement::isSystemPreviewImage const):
(WebCore::HTMLImageElement::cloneElementWithoutAttributesAndChildren):
* Source/WebCore/html/HTMLImageLoader.cpp:
(WebCore::HTMLImageLoader::notifyFinished):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::willDispatchEvent):
* Source/WebCore/html/HTMLLIElement.cpp:
(WebCore::HTMLLIElement::didAttachRenderers):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::isEventTargetedAtInteractiveDescendants const):
(WebCore::HTMLLabelElement::defaultEventHandler):
* Source/WebCore/html/HTMLLegendElement.cpp:
(WebCore::HTMLLegendElement::form const):
* Source/WebCore/html/HTMLMaybeFormAssociatedCustomElement.h:
(isType):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::bestMediaElementForRemoteControls):
(WebCore::HTMLMediaElement::updateActiveTextTrackCues):
(WebCore::HTMLMediaElement::textTrackRemoveCue):
(WebCore::HTMLMediaElement::setReadyState):
(WebCore::HTMLMediaElement::mediaPlayerSizeChanged):
(WebCore::HTMLMediaElement::mediaPlayerRenderingCanBeAccelerated):
(WebCore::HTMLMediaElement::enterFullscreen):
(WebCore::HTMLMediaElement::exitFullscreen):
(WebCore::HTMLMediaElement::willBecomeFullscreenElement):
(WebCore::HTMLMediaElement::mediaState const):
(WebCore::HTMLMediaElement::isVideoTooSmallForInlinePlayback):
* Source/WebCore/html/HTMLMediaElement.h:
(isType):
* Source/WebCore/html/HTMLMeterElement.cpp:
(WebCore::HTMLMeterElement::renderMeter const):
* Source/WebCore/html/HTMLNameCollection.cpp:
(WebCore::isObjectElementForDocumentNameCollection):
* Source/WebCore/html/HTMLObjectElement.cpp:
(WebCore::preventsParentObjectFromExposure):
(WebCore::HTMLObjectElement::updateExposedState):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::ownerSelectElement const):
(WebCore::HTMLOptionElement::isDisabledFormControl const):
(WebCore::HTMLOptionElement::collectOptionInnerText const):
* Source/WebCore/html/HTMLPictureElement.cpp:
(WebCore::HTMLPictureElement::isSystemPreviewImage):
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::defaultEventHandler):
(WebCore::HTMLPlugInElement::supportsFocus const):
(WebCore::HTMLPlugInElement::setReplacement):
(WebCore::HTMLPlugInElement::isReplacementObscured):
* Source/WebCore/html/HTMLPlugInImageElement.cpp:
(WebCore::HTMLPlugInImageElement::didAttachRenderers):
* Source/WebCore/html/HTMLPlugInImageElement.h:
(isType):
* Source/WebCore/html/HTMLProgressElement.cpp:
(WebCore::HTMLProgressElement::renderProgress const):
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::optionToSelectFromChildChangeScope):
(WebCore::HTMLSelectElement::setLength):
(WebCore::HTMLSelectElement::nextSelectableListIndexPageAway const):
(WebCore::HTMLSelectElement::saveLastSelection):
(WebCore::HTMLSelectElement::setActiveSelectionAnchorIndex):
(WebCore::HTMLSelectElement::updateListBoxSelection):
(WebCore::HTMLSelectElement::listBoxOnChange):
(WebCore::HTMLSelectElement::scrollToSelection):
(WebCore::HTMLSelectElement::setOptionsChangedOnRenderer):
(WebCore::HTMLSelectElement::recalcListItems const):
(WebCore::HTMLSelectElement::selectedIndex const):
(WebCore::HTMLSelectElement::selectOption):
(WebCore::HTMLSelectElement::deselectItemsWithoutValidation):
(WebCore::HTMLSelectElement::saveFormControlState const):
(WebCore::HTMLSelectElement::searchOptionsForValue const):
(WebCore::HTMLSelectElement::restoreFormControlState):
(WebCore::HTMLSelectElement::appendFormData):
(WebCore::HTMLSelectElement::reset):
(WebCore::HTMLSelectElement::platformHandleKeydownEvent):
(WebCore::HTMLSelectElement::menuListDefaultEventHandler):
(WebCore::HTMLSelectElement::updateSelectedState):
(WebCore::HTMLSelectElement::listBoxDefaultEventHandler):
(WebCore::HTMLSelectElement::defaultEventHandler):
(WebCore::HTMLSelectElement::lastSelectedListIndex const):
(WebCore::HTMLSelectElement::optionAtIndex const):
(WebCore::HTMLSelectElement::accessKeySetSelectedIndex):

Canonical link: https://commits.webkit.org/271033@main
